src/exchange.c

```diff
@@ -231,6 +231,17 @@ enum wps_result do_wps_exchange()
 			set_timeout_is_nack(0);
 
 			ret_val = KEY_REJECTED;
+
+			/* Check the reason code for the received NACK message */
+			if (get_nack_reason() == MESSAGE_TIMEOUT) {
+				ret_val = UNKNOWN_ERROR;
+				cprintf(WARNING, "[!] WARNING: Potential FAKE NACK!\n");
+			}
+			/* Got NACK instead of an M5 message, when cracking second half */
+			else if (last_msg == M3 && get_key_status() == KEY2_WIP) {
```

Comment:
By logic, yet missing the checking of `get_pin_string_mode()`

Reply:
sorry, could you elaborate what you mean ?

Reply:
When we specified a not valid pin or empty pin, the first half pin and second half pin are irrelevants and consequently the status KEY2_WIP is invalid/fake after received M5, for to reinforce not to enter the block it is necessary `get_pin_string_mode()`

Reply:
i see. can you push a commit addressing that ?

Reply:
I force-pushed the last commit 387678d with that